File: 1. 단계별로_풀어보기/30~39/33. 트리/1. 11725.py
import sys

sys.setrecursionlimit(100000)

input = sys.stdin.readline
output = sys.stdout.write

N = int(input())
treeInfoArr = list(list(map(int, input().split())) for i in range(N - 1))

connArr = list([] for i in range(N + 1))
parentArr = list(-1 for i in range(N + 1))

parentArr[1] = 1
for a, b in treeInfoArr:
  connArr[a].append(b)
  connArr[b].append(a)

def DFS(now):
  for next in connArr[now]:
    if parentArr[next] == -1:
      parentArr[next] = now
      DFS(next)

DFS(1)

output(f"{'\n'.join(str(num) for num in parentArr[2:])}")

File: 1. 단계별로_풀어보기/33. 트리/1. 11725.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const N = +inputArr[0]
const treeInfoArr = inputArr.slice(1).map(row => row.trim().split(' ').map(Number))

// connArr[i]: i 와 연결된 정점들의 목록
const connArr = Array.from({length: N + 1}, () => [])

treeInfoArr.forEach(info => {
  const [start, end] = info
  connArr[start].push(end)
  connArr[end].push(start)
})

// parentArr[i]: i 의 부모
const parentArr = Array.from({length: N + 1}).fill(-1)
parentArr[1] = 1

// BFS 용 방문큐. 배열로 선언해도 무방하다
const visitQueue = Array.from({length: N + 1}).fill(0)
let nowIdx = 0
let insertIdx = 0

visitQueue[insertIdx++] = 1

while (nowIdx < insertIdx) {
  const now = visitQueue[nowIdx++]

  connArr[now].forEach(next => {
    if (parentArr[next] == -1) {
      parentArr[next] = now
      visitQueue[insertIdx++] = next
    }
  })
}

// 출력 변수. 이거 안하면 시간초과 생길 수 있음
let resultStr = ''

parentArr.slice(2).forEach(parent => (resultStr += `${parent}\n`))

process.stdout.write(resultStr)
